eclair-core/src/main/scala/fr/acinq/eclair/payment/send/MultiPartPaymentLifecycle.scala

```diff
@@ -72,108 +66,91 @@ class MultiPartPaymentLifecycle(nodeParams: NodeParams, cfg: SendPaymentConfig,
+      log.debug("sending {} with maximum fee {}", r.totalAmount, maxFee)
+      val d = PaymentProgress(sender, r, r.maxAttempts, Map.empty, Set.empty, Set.empty, Nil)
+      router ! d.createRouteRequest(nodeParams, r.totalAmount, maxFee, routeParams)
+      goto(WAIT_FOR_ROUTES) using d
   }
 
-  when(WAIT_FOR_NETWORK_STATS) {
-    case Event(s: GetNetworkStatsResponse, d: WaitingForNetworkStats) =>
-      log.debug("network stats: {}", s.stats.map(_.capacity))
-      // If we don't have network stats it's ok, we'll use data about our local channels instead.
-      // We tell the router to compute those stats though: in case our payment attempt fails, they will be available for
-      // another payment attempt.
-      if (s.stats.isEmpty) {
-        router ! TickComputeNetworkStats
-      }
-      relayer ! GetOutgoingChannels()
-      goto(WAIT_FOR_CHANNEL_BALANCES) using WaitingForChannelBalances(d.sender, d.request, s.stats)
-  }
-
-  when(WAIT_FOR_CHANNEL_BALANCES) {
-    case Event(OutgoingChannels(channels), d: WaitingForChannelBalances) =>
-      log.debug("trying to send {} with local channels: {}", d.request.totalAmount, channels.map(_.toUsableBalance).mkString(","))
-      val (remaining, payments) = splitPayment(nodeParams, d.request.totalAmount, channels, d.networkStats, d.request, randomize = false)
-      if (remaining > 0.msat) {
-        log.warning(s"cannot send ${d.request.totalAmount} with our current balance")
-        Metrics.PaymentError.withTag(Tags.Failure, Tags.FailureType(LocalFailure(Nil, PaymentError.BalanceTooLow)))
-        goto(PAYMENT_ABORTED) using PaymentAborted(d.sender, d.request, LocalFailure(Nil, PaymentError.BalanceTooLow) :: Nil, Set.empty)
-      } else {
-        val pending = setFees(d.request.routeParams, payments, payments.size)
+  when(WAIT_FOR_ROUTES) {
+    case Event(RouteResponse(routes), d: PaymentProgress) =>
+      log.info("{} routes found (attempt={}/{})", routes.length, d.request.maxAttempts - d.remainingAttempts + 1, d.request.maxAttempts)
+      val (toSend, maxFee) = remainingToSend(nodeParams, d.request, d.pending.values)
+      if (routes.map(_.amount).sum == toSend) {
+        val childPayments = routes.map(route => (UUID.randomUUID(), route)).toMap
         Kamon.runWithContextEntry(parentPaymentIdKey, cfg.parentId) {
           Kamon.runWithSpan(span, finishSpan = true) {
-            pending.foreach { case (childId, payment) => spawnChildPaymentFsm(childId) ! payment }
+            childPayments.foreach { case (childId, route) => spawnChildPaymentFsm(childId) ! createChildPayment(route, d.request) }
           }
         }
-        goto(PAYMENT_IN_PROGRESS) using PaymentProgress(d.sender, d.request, d.networkStats, channels.length, 0 msat, d.request.maxAttempts - 1, pending, Set.empty, Nil)
+        goto(PAYMENT_IN_PROGRESS) using d.copy(remainingAttempts = (d.remainingAttempts - 1).max(0), pending = d.pending ++ childPayments)
```

Comment:
Yes, since the HTLC fan-out can vary greatly (because we can either split in many parts or not), the way I see it each attempt should map to a run of the path-finding algorithm to try a different set of routes (with a small subtlety when we discard routes because of a racing failure and don't want to count that as an attempt).